File: RecyclerView/app/src/main/java/com/xyz/recyclerview/MainActivity.java
package com.xyz.recyclerview;

import android.os.Bundle;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements ItemClickListener {

    private RecyclerView recyclerView;
    private List<Animal> menuList;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        recyclerView = findViewById(R.id.rvMenu);
        buildRecyclerData();
        setRecyclerAdapter();
    }

    private void setRecyclerAdapter() {
        AnimalAdapter menuAdapter = new AnimalAdapter(menuList, this);
//        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
//        GridLayoutManager layoutManager = new GridLayoutManager(this, 4);
        StaggeredGridLayoutManager layoutManager = new StaggeredGridLayoutManager(4, LinearLayoutManager.VERTICAL);
        recyclerView.setLayoutManager(layoutManager);
        /*
        Pass the data to the adapter class
         */
        recyclerView.setAdapter(menuAdapter);
    }

    private void buildRecyclerData() {
        menuList = new ArrayList<>();
        for (int i = 0; i < 100; i++) {

        }
    }

    @Override
    public void onItemClicked(Animal menu, int position) {
        Toast.makeText(this, "Item clicked at position " + (position + 1), Toast.LENGTH_SHORT).show();
    }
}

File: Retrofit-Kotlin/app/src/main/java/com/xyz/retrofit_kotlin/MainActivity.kt
package com.xyz.retrofit_kotlin

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import kotlinx.android.synthetic.main.activity_main.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        callApi()
    }

    private fun callApi() {
        val apiClient = Network.getInstance().create(ApiClient::class.java)
        val call = apiClient.getPosts()
        call.enqueue(object : Callback<List<ResponseModel>> {
            override fun onResponse(
                call: Call<List<ResponseModel>>,
                response: Response<List<ResponseModel>>
            ) {
                if (response.body() != null) {
                    response.body()?.let {
                        buildData(it)
                    }

                }
            }

            override fun onFailure(call: Call<List<ResponseModel>>, t: Throwable) {
            }

        })
    }

    private fun buildData(list: List<ResponseModel>) {
        val layoutManager = LinearLayoutManager(this@MainActivity)
        recyclerview.layoutManager = layoutManager
        val postAdapter = PostAdapter(list)
        recyclerview.adapter = postAdapter
    }
}